common/flatpak-transaction.c

```diff
@@ -2431,6 +2658,262 @@ resolve_all_ops (FlatpakTransaction *self,
+ * cancel the ongoing authentication operation.
+ *
+ * This is useful for example if you're showing an authenticaion
+ * window with a browser, but the user closed it before it was finished.
+ */
+void
+flatpak_transaction_abort_webflow (FlatpakTransaction *self,
+                                   guint id)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_autoptr(GError) local_error = NULL;
+
+  if (priv->active_webflow_id == id)
+    {
+      RequestData *data = priv->active_webflow;
+
+      g_assert (data != NULL);
+      priv->active_webflow_id = 0;
+
+      if (!data->done)
+        {
+          if (!flatpak_authenticator_request_call_close_sync (data->request, NULL, &local_error))
+            g_debug ("Failed to close auth request: %s", local_error->message);
+        }
+    }
+}
+
+static gboolean
+request_tokens_for_remote (FlatpakTransaction *self,
+                           const char         *remote,
+                           GList              *ops,
+                           GCancellable       *cancellable,
+                           GError            **error)
+{
+  FlatpakTransactionPrivate *priv = flatpak_transaction_get_instance_private (self);
+  g_autoptr(GString) refs_as_str = g_string_new ("");
+  GList *l;
+  g_autoptr(AutoFlatpakAuthenticatorRequest) request = NULL;
+  g_autoptr(AutoFlatpakAuthenticator) authenticator = NULL;
+  g_autoptr(GMainContextPopDefault) context = NULL;
+  RequestData data = { self, remote };
+  g_autoptr(GVariant) tokens = NULL;
+  g_autoptr(GVariant) results = NULL;
+  g_autoptr(GVariant) refs = NULL;
+  GVariantBuilder refs_builder;
+
+  g_variant_builder_init (&refs_builder, G_VARIANT_TYPE ("a(si)"));
+
+  for (l = ops; l != NULL; l = l->next)
+    {
+      FlatpakTransactionOperation *op = l->data;
+      g_variant_builder_add (&refs_builder, "(si)", op->ref, (gint32)op->token_type);
+      g_string_append_printf (refs_as_str, "(%s, %d)", op->ref, op->token_type);
+      if (l->next != NULL)
+        g_string_append (refs_as_str, ", ");
+    }
+
+  g_debug ("Requesting tokens for remote %s: %s", remote, refs_as_str->str);
+  refs = g_variant_ref_sink (g_variant_builder_end (&refs_builder));
+
+  context = flatpak_main_context_new_default ();
+
+  authenticator = flatpak_auth_new_for_remote (priv->dir, remote, cancellable, error);
+  if (authenticator == NULL)
+    return FALSE;
+
+  request = flatpak_auth_create_request (authenticator, cancellable, error);
+  if (request == NULL)
+    return FALSE;
+
+  g_signal_connect (request, "webflow", (GCallback)request_tokens_webflow, &data);
+  g_signal_connect (request, "webflow-done", (GCallback)request_tokens_webflow_done, &data);
+  g_signal_connect (request, "response", (GCallback)request_tokens_response, &data);
+
+  priv->active_webflow = &data;
+
+  data.request = request;
+  if (!flatpak_auth_request_ref_tokens (authenticator, request, remote, refs, cancellable, error))
+    return FALSE;
```

Comment:
Here you are returning without resetting priv->active_webflow

Reply:
Not sure what you mean, this is an async call and its reset a few lines below after iterating the mainloop, no?